marconi/src/Marconi/Index/New/ScriptTx.hs

```diff
@@ -0,0 +1,418 @@
+
+instance SQL.ToRow ScriptTxRow where
+  toRow o = [ SQL.toField $ scriptAddress o
+            , SQL.toField $ txCbor o
+            , SQL.toField $ txSlot o ]
+
+deriving anyclass instance SQL.FromRow ScriptTxRow
+
+-- * Indexer
+
+type Query = StorableQuery ScriptTxHandle
+type Result = StorableResult ScriptTxHandle
+
+toUpdate
+  :: forall era . C.IsCardanoEra era
+  => [C.Tx era]
+  -> SlotNo
+  -> StorableEvent ScriptTxHandle
+toUpdate txs = ScriptTxEvent txScripts'
+  where
+    txScripts' = map (\tx -> (TxCbor $ C.serialiseToCBOR tx, getTxScripts tx)) txs
+
+getTxBodyScripts :: forall era . C.TxBody era -> [StorableQuery ScriptTxHandle]
+getTxBodyScripts body = let
+    hashesMaybe :: [Maybe C.ScriptHash]
+    hashesMaybe = case body of
+      Shelley.ShelleyTxBody shelleyBasedEra _ scripts _ _ _ ->
+        flip map scripts $ \script ->
+          case fromShelleyBasedScript shelleyBasedEra script of
+            Shelley.ScriptInEra _ script' -> Just $ C.hashScript script'
+      _ -> [] -- Byron transactions have no scripts
+    hashes = catMaybes hashesMaybe :: [Shelley.ScriptHash]
+  in map ScriptTxAddress hashes
+
+getTxScripts :: forall era . C.Tx era -> [StorableQuery ScriptTxHandle]
+getTxScripts (C.Tx txBody _ws) = getTxBodyScripts txBody
+
+{- Now that all connected data types have been defined, we go on to implement some
+   of the type classes required for information storage and retreival. -}
+
+instance Buffered ScriptTxHandle where
+  {- The data is buffered in memory. When the memory buffer is filled, we need to store
+     it on disk. -}
+  persistToStorage
+    :: Foldable f
+    => f (StorableEvent ScriptTxHandle)
+    -> ScriptTxHandle
+    -> IO ScriptTxHandle
+  persistToStorage es h = do
+    let rows = foldl' (\ea e -> ea ++ flatten e) [] es
+        c    = hdlConnection h
+    SQL.execute_ c "BEGIN"
+    forM_ rows $
+      SQL.execute c "INSERT INTO script_transactions (scriptAddress, txCbor, slotNo) VALUES (?, ?, ?)"
+    SQL.execute_ c "COMMIT"
+    pure h
+
+    where
+      flatten :: StorableEvent ScriptTxHandle -> [ScriptTxRow]
+      flatten (ScriptTxEvent txs sl) = do
+        (tx, scriptAddrs) <- txs
+        addr <- scriptAddrs
+        pure $ ScriptTxRow { scriptAddress = addr
+                           , txCbor        = tx
+                           , txSlot        = sl
+                           }
+
+  {- We want to potentially store data in two formats. The first one is similar (if
+     not identical) to the format of data stored in memory; it should contain information
+     that allows knowing at which point the data was generated.
+
+     We use this first format to support rollbacks for disk data. The second format,
+     which is not always necessary and does not have any predetermined structure,
+     should be thought of as an aggregate of the previously produced events.
+
+     For this indexer we don't really need an aggregate, so our "aggregate" has almost the same
```

Comment:
If we did need an aggregate, where would we define it?

Reply:
A separate database table with whatever structure is efficient for your queries.

Reply:
I can talk about an example when implementing a UTXOs aggregate. Do you think that would be a good idea?

Reply:
I think the best would be to actually write an indexer that implements a simple aggregate. For example, an indexer which aggregates the `Value` for each `Address` (like in an accounting model). I could create a story for this?